processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/TypeSelector.java

```diff
@@ -34,17 +37,154 @@
  */
 public class TypeSelector implements MethodSelector {
 
+    private TypeFactory typeFactory;
+
+    public TypeSelector(TypeFactory typeFactory) {
+        this.typeFactory = typeFactory;
+    }
+
     @Override
-    public <T extends Method> List<T> getMatchingMethods(Method mappingMethod, List<T> methods,
-                                                         Type sourceType, Type targetType,
-                                                         SelectionCriteria criteria) {
-
-        List<T> result = new ArrayList<T>();
-        for ( T method : methods ) {
-            if ( !method.isLifecycleCallbackMethod() && method.matches( Arrays.asList( sourceType ), targetType ) ) {
-                result.add( method );
+    public <T extends Method> List<MethodSelection<T>> getMatchingMethods(Method mappingMethod,
+                                                                          List<MethodSelection<T>> methods,
+                                                                          List<Type> sourceTypes, Type targetType,
+                                                                          SelectionCriteria criteria) {
+
+        if ( methods.isEmpty() ) {
+            return methods;
+        }
+
+        List<MethodSelection<T>> result = new ArrayList<MethodSelection<T>>();
+
+        List<Parameter> availableParams;
```

Comment:
I'm trying to follow what is happening here.. Could you add some comments..? What is relevant for factories? What is relevant for lifecycle methods? Or is this going to change when `Context` is added?